### PR TITLE
net: sockets: Implement sendto() and recvfrom()

### PR DESCRIPTION
**Pull request by askawu**
Support sendto() and recvfrom() for datagram socket.

### CODE REVIEW

File: subsys/net/lib/sockets/sockets.c [hunk 1]
```diff
@@ -460,3 +460,11 @@ int zsock_poll(struct zsock_pollfd *fds, int nfds, int timeout)
 
 	return ret;
 }
+
+int zsock_inet_pton(sa_family_t family, const char *src, void *dst)
+{
+	if (net_addr_pton(family, src, dst) == 0)
```
Comment (tbursztyka) [inline subsys/net/lib/sockets/sockets.c]: add {} always, even on one liners. this is a zephyr style requirement

File: subsys/net/lib/sockets/sockets.c [hunk 2]
```diff
@@ -209,45 +210,100 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
 	return POINTER_TO_INT(ctx);
 }
 
+static struct net_pkt* get_tx_pkt(struct net_context *ctx,
```
Comment (tbursztyka) [inline subsys/net/lib/sockets/sockets.c]: this is not a getter function anymore as you are appending data as well in the process, you'll have to rename this one.

File: subsys/net/lib/sockets/sockets.c [hunk 3]
```diff
@@ -209,45 +210,100 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
+}
+
+ssize_t zsock_sendto(int sock, const void *buf, size_t len, int flags,
+		     const struct sockaddr *dest_addr, socklen_t addrlen)
+{
+	ARG_UNUSED(flags);
+	int err;
+	struct net_pkt *send_pkt;
+	s32_t timeout = K_FOREVER;
+	struct net_context *ctx = INT_TO_POINTER(sock);
```
Comment (tbursztyka) [inline subsys/net/lib/sockets/sockets.c]: reverse the order of all the 5 lines above.

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -209,45 +210,100 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
+	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	/* auto bind if necessary */
+	if (!net_context_is_bound(ctx)) {
+		struct sockaddr src_addr;
+		memset(&src_addr, 0, sizeof(src_addr));
```
Comment (tbursztyka) [inline subsys/net/lib/sockets/sockets.c]: lines break are cheap and help for code readability, add empty lines after variable declaration.

File: subsys/net/ip/net_pkt.c [hunk 5]
```diff
@@ -1647,6 +1647,58 @@ void net_pkt_get_info(struct k_mem_slab **rx,
...
+int net_pkt_get_src_addr(struct net_pkt *pkt, struct sockaddr *addr,
+			 socklen_t *addrlen)
+{
+	struct net_context *ctx;
+	enum net_ip_protocol proto;
+	sa_family_t family;
+	u16_t port = 0;
+
+	if (!addr || !pkt || !addrlen)
+		return -EINVAL;
```
Comment (tbursztyka) [inline subsys/net/ip/net_pkt.c]: {} needed

File: subsys/net/ip/net_context.c [hunk 6]
```diff
@@ -2458,6 +2458,29 @@ int net_context_update_recv_wnd(struct net_context *context,
 	return -EPROTOTYPE;
 #endif
 }
+
+bool net_context_is_bound(struct net_context *context)
+{
+	NET_ASSERT(context);
+
+#if defined(CONFIG_NET_IPV6)
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c]: use IS_ENABLED with this CONFIG option, in the test right below.

Do the same for ipv4 also

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Review by pfalcon — CHANGES_REQUESTED** [review]
As a quick reply, this seems more complex than intuitively expected.

My intuition can be wrong of course, but one issue I see is trying to resolve "default local bind for UDP" issue on BSD sockets level. This isn't good approach. We should resolve as many as possible issues on net_context (i.e. native Z API level), to make sure that it, and any levels above it benefit from it.

I introduced a framework for default local binding in #817, and applied it to net_context_connect(). It was on my TODO to apply it to UDP use cases which don't require explicit connect(), but that slipped.

So, I'd recommend to use the same approach and already introduced functions from #817 to handle unbound UDP context in net_context_send(to)() and net_context_recv().

Just as send/recv operations are usually on critical path, I'd suggest calling bind_default() only on UDP paths (as TCP should be already bound in connect()).

The above would be a separate PR. And actually, I'd suggest making  inet_pton() fix (good catch!) its own PR too.

**Review by jukkar — COMMENTED** [review]
I do not have anything extra, Tomasz pretty much commented everything already.

**Comment by jukkar** [comment]
> I introduced a framework for default local binding in #817, and applied it to net_context_connect(). It was on my TODO to apply it to UDP use cases which don't require explicit connect(), but that slipped.

Indeed, good point.

**Comment by askawu** [comment]
@tbursztyka , I will correct the coding style problem. Thanks
@pfalcon , at the beginning, I tried to auto bind the local address by bind_default() in net context and it worked. But it become complicated when registering the recv callback before sendto(). Anyway, I will make the inet_pton patch as a new PR and see if I can make auto bind work for UDP in net context.

**Comment by pfalcon** [comment]
@askawu: Any updates here? Would you like me to give a try to resolving auto-bind issue for UDP sockets?

**Comment by askawu** [comment]
@pfalcon, I updated the patch but I forgot to update the PR.  Can you help to check if a81cbe2 makes sense to you? bind_default() is used to auto bind UDP soccket.

File: subsys/net/ip/net_context.c [hunk 7]
```diff
@@ -2096,6 +2096,11 @@ static int sendto(struct net_pkt *pkt,
 		return -EBADF;
 	}
 
+	ret = bind_default(context);
```
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: I would definitely suggest to call this only for UDP case, there's no need to do an extra call for TCP case. There're 2 ways to do that:

1. If we care about source addr to be bound for offloading case, then just add bind_default() block wrapped in `#if defined(CONFIG_NET_UDP)` after existing `#if defined(CONFIG_NET_TCP)` right below.
2. If we can cheat on offloading case, can reuse existing `if (net_context_get_ip_proto(context) == IPPROTO_UDP)` far below.

@jukkar : What do you think?

Comment (jukkar) [reply subsys/net/ip/net_context.c]: Sure, makes sense.

Comment (pfalcon) [reply subsys/net/ip/net_context.c]: @jukkar: Thanks for the reply, but it's "either/or" type of question (either way 1 above, or way 2) ;-). Please suggest which way to go.

File: subsys/net/ip/net_pkt.c [hunk 8]
```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
...
+
+		port = hdr->src_port;
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		struct sockaddr_in6 *addr6 = net_sin6(addr);
+
+		if (*addrlen < sizeof(struct sockaddr_in6)) {
```
Comment (pfalcon) [inline subsys/net/ip/net_pkt.c]: What about returning the address len in *addrlen? If you didn't intend to do that, why is it passed by pointer?

File: include/net/net_pkt.h [hunk 9]
```diff
@@ -1346,6 +1346,19 @@ void net_pkt_get_info(struct k_mem_slab **rx,
 		      struct net_buf_pool **rx_data,
 		      struct net_buf_pool **tx_data);
 
+/**
+ * @brief Get source socket address.
+ *
+ * @param pkt Nework packet
+ * @param addr Source socket address
+ * @param addrlen The length of source socket address
```
Comment (pfalcon) [inline include/net/net_pkt.h]: Once semantics of *addrlen figure out (comment below), this docstring should describe what gets there on input, what on output.

File: subsys/net/lib/sockets/sockets.c [hunk 10]
```diff
@@ -209,45 +210,92 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
+	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	if (err < 0) {
+		net_pkt_unref(send_pkt);
+		errno = -err;
+		return -1;
 	}
 
-	if (len > max_len) {
-		len = max_len;
+	return len;
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: This doesn't appear to be correct.

File: subsys/net/lib/sockets/sockets.c [hunk 11]
```diff
@@ -209,45 +210,92 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
+}
+
+ssize_t zsock_sendto(int sock, const void *buf, size_t len, int flags,
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: I would do it differently - zsock_send() should just call to zsock_sendto() with NULL dest address. Given that there's issue with length returned (previous comment), you might need to it like that.

File: subsys/net/lib/sockets/sockets.c [hunk 12]
```diff
@@ -209,45 +210,92 @@ int zsock_accept(int sock, struct sockaddr *addr, socklen_t *addrlen)
...
-	if (!len) {
-		net_pkt_unref(send_pkt);
+	send_pkt = prepare_tx_pkt(ctx, buf, len, timeout);
+	if (!send_pkt) {
 		errno = EAGAIN;
 		return -1;
 	}
 
-	err = net_context_send(send_pkt, /*cb*/NULL, timeout, NULL, NULL);
+	SET_ERRNO(net_context_recv(ctx, zsock_received_cb, K_NO_WAIT, NULL));
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Why is this line here? (And not in zsock_send()? zsock_send() and zsock_sendto() should behave the same, and as noted previously, the way to ensure that is make one tail-call another).

File: subsys/net/lib/sockets/sockets.c [hunk 13]
```diff
@@ -332,10 +382,17 @@ static inline ssize_t zsock_recv_stream(struct net_context *ctx, void *buf, size
 
 ssize_t zsock_recv(int sock, void *buf, size_t max_len, int flags)
 {
+	return zsock_recvfrom(sock, buf, max_len, flags, NULL, NULL);
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Yes, this is right. zsock_send() should be done like this.

File: subsys/net/lib/sockets/sockets.c [hunk 14]
```diff
@@ -143,11 +143,12 @@ static void zsock_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 	/* Normal packet */
 	net_pkt_set_eof(pkt, false);
 
-	/* We don't care about packet header, so get rid of it asap */
-	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
-	net_buf_pull(pkt->frags, header_len);
-
 	if (net_context_get_type(ctx) == SOCK_STREAM) {
+		/* TCP: we don't care about packet header, get rid of it asap.
```
Comment (pfalcon) [inline subsys/net/lib/sockets/sockets.c]: Why this conditionalizing, leading to code duplication? If we (may) need a packet header, let's keep it here, and consistently remove in zsock_recv[from]().

**Comment by pfalcon** [comment]
@askawu: Thanks for refactors so far, and sorry for the delay with the review.

Well, unfortunately it's still not there IMHO, some issues need fixing, and some things can be redone in an easier/cleaner way. Per-commit comments:

> net: pkt: Add net_pkt_get_src_addr() 

I was surprised that such a function needs to be introduced. How did we do it before? I checked echo_server and yeah, it picks src addr (to be dst addr on reply) in an adhoc way. @jukkar , how do you deal with it in http server? Would the func above be useful there, would you refactor to use it?

> net: context: Bind default address for UDP

My usual suggestion goes here: please split unrelated patches to a standalobe PRs to allow for quicker review cycle. I also suggest to optimize TCP case (for which we know that context is already bound). There's a pending question to @jukkar how to best do that.

> net: sockets: Implement sendto() and recvfrom()

zsock_recvfrom() is implemented like it should be - by making zsock_recv() to dispatch to it. There's still a code duplication to optimize though.

But instead of doing the same for zsock_sendto() and zsock_send(), it's done in adhoc way, which in particular leads to a bug with return length. Please redo the same way as  zsock_recvfrom()/zsock_recv(). Besides being easier and cleaner that way, that's also definitely a requirement for socket offloading, I hope @GAnthony can second on that.

Thanks!

File: subsys/net/ip/net_context.c (hunk 7 above)
Comment (pfalcon) [reply subsys/net/ip/net_context.c]: @GAnthony : Maybe you can suggest here something too.

Comment (askawu) [reply subsys/net/ip/net_context.c]: I prefer the second one.

File: subsys/net/ip/net_pkt.c (hunk 8 above)
Comment (askawu) [reply subsys/net/ip/net_pkt.c]: Yes, the address size isn't necessary to be a pointer.

File: subsys/net/lib/sockets/sockets.c (hunk 12 above)
Comment (askawu) [reply subsys/net/lib/sockets/sockets.c]: My assumption is, for TCP case, zsock_connect() will be called first and callback will be registered at that time.  So I only add it in zsock_sendto().

If we make zsock_send() and zsock_send() in adhoc way, any idea to prevent the registering the callback twice? Is there a proper way in net context to check if the callback has been registered?

File: subsys/net/ip/net_pkt.c (hunk 8 above)
Comment (pfalcon) [reply subsys/net/ip/net_pkt.c]: Hmm, I'd say "it's useful to return the actual addrlen". Because otherwise, we probably don't even need to pass it in - struct sockaddr can accommodate both IPv4 and IPv6 addresses.

File: subsys/net/lib/sockets/sockets.c (hunk 14 above)
Comment (askawu) [reply subsys/net/lib/sockets/sockets.c]: Initially, I prefer to keep the packet header for both cases. But I found zsock_recv_stream() needs to be modified a lot because packet header isn't dropped. I will modify this to prefer consistency.

**Comment by askawu** [comment]
@pfalcon, Thanks for the comments. I will think about how to make zsock_send() and zsock_sendto() easier.
@jukkar, Can you help to review 8296c76? I would like to know if there's a better way to get the src address from the packet.

File: subsys/net/lib/sockets/sockets.c (hunk 14 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: Hmm, you may be right - and we can't return src addr for an arbitrary TCP recv anyway (only for recv() which starts to consume a new packet). Well, that's why it's useful to know why something was implemented this or that way, so having extra info in the commit messages or PR comments is helpful. And yeah, I'd appreciate your giving another thought to trying to avoid that code duplication - if it will lead to only more code being added, sure, it's ok to leave it like that.

File: subsys/net/lib/sockets/sockets.c (hunk 12 above)
Comment (pfalcon) [reply subsys/net/lib/sockets/sockets.c]: I see. Yep, there're some non-obvious points which only come to attention when you dig deeper into things. (Which may hint that there should be a code comment about it!)

> If we make zsock_send() and zsock_send() in adhoc way, any idea to prevent the registering the callback twice? Is there a proper way in net context to check if the callback has been registered?

Well, so you already register a callback here multiple times - each call to sendto() will lead to callback being registered. If that works, then testing for UDP and setting it in send() should also work.

But overall, that's a similar problem to bind_default(). I'd say, making send() to tail-cal to sendto() is more priority than avoiding duplicate setting if recv callback. Let's try to optimize it as we can now (i.e. set it in send*() only for UDP case), and I already accept that we'll need a much later pass over the socket code to see how we can optimize things.

File: subsys/net/ip/net_pkt.c (hunk 8 above)
Comment (askawu) [reply subsys/net/ip/net_pkt.c]: It's used to prevent out of bound access. For example, the address is IPv6 but a sockaddr_in variable is passed in. I know it's rare but the caller might make a mistake. In such case, an error will be returned, so I probably don't need to return the actual addrlen.

Comment (pfalcon) [reply subsys/net/ip/net_pkt.c]: Ok, I assume you considered various usecases for this function, so sounds good. (And we can change it later if found useful.)

File: subsys/net/ip/net_pkt.c [hunk 15]
```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
...
+	sa_family_t family;
+	u16_t port = 0;
+
+	if (!addr || !pkt) {
+		return -EINVAL;
+	}
+
+	ctx = pkt->context;
+	proto = net_context_get_ip_proto(ctx);
+	family = net_context_get_family(ctx);
```
Comment (jukkar) [inline subsys/net/ip/net_pkt.c]: The pkt might not have context specified so this is error prone.
Family can be get using net_pkt_family(pkt) instead. For protocol, we could get that information from IP header, or we could store it to net_pkt header when packet is received. We have been trying to avoid increasing the size of net_pkt so first option could be done here.

So depending on family, the UDP/TCP information can be get like this
`proto = NET_IPV6_HDR(pkt)->nexthdr;`
or
`proto = NET_IPV4_HDR(pkt)->proto;`

File: subsys/net/ip/net_pkt.c [hunk 16]
```diff
@@ -1654,6 +1654,59 @@ void net_pkt_get_info(struct k_mem_slab **rx,
...
+		port = hdr->src_port;
+	} else if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+		struct net_udp_hdr *hdr = net_pkt_udp_data(pkt);
+
+		port = hdr->src_port;
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
```
Comment (jukkar) [inline subsys/net/ip/net_pkt.c]: If you move this block before checking UDP/TCP, then you could add the proto resolving here.

**Review by jukkar — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
> I was surprised that such a function needs to be introduced. How did we do it before? I checked echo_server and yeah, it picks src addr (to be dst addr on reply) in an adhoc way. @jukkar , how do you deal with it in http server? Would the func above be useful there, would you refactor to use it?

That functions sounds useful to have, +1 for that.

**Comment by jukkar** [comment]
> I also suggest to optimize TCP case (for which we know that context is already bound). There's a pending question to @jukkar how to best do that.

I must have missed this question, could you elaborate what was this question about?

**Comment by askawu** [comment]
Update:
- net_pkt_get_src_addr() should use the family and protocol from packet header instead of the value in net context. 
- Bind default address and port only for UDP
- Refactor zsock_send() and zsock_sendto() to remove the duplicated code.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by pfalcon — APPROVED** [review]
LGTM, thanks for all the refactoring!

@tbursztyka : Your comments also should have been addressed, can you please have a look?

**Review by tbursztyka — APPROVED** [review]
